src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1682,14 +1732,73 @@ PlanRouterQuery(Query *originalQuery, RelationRestrictionContext *restrictionCon
 
 
 /*
+ * MultiShardModifyQuerySupported checks whether the query and pruned relation
+ * shard list is supported. It is supported if there is single table in the
+ * relation shard list and query doesn't contain any volatile function.
+ */
+static DeferredErrorMessage *
+MultiShardModifyQuerySupported(List *relationShardList)
+{
+	DeferredErrorMessage *planningError = NULL;
+
+	if (MultipleDistributedTableList(relationShardList))
+	{
+		ereport(DEBUG2, (errmsg("Support only single distributed table with multi"
+								" shard modify commands")));
+
+		planningError = DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+									  "support only single distributed table with "
+									  "multi shard modify commands",
+									  NULL, NULL);
+		return planningError;
+	}
+
+	return NULL;
+}
+
+
+/*
+ * MultipleDistributedTableList checks whether the pruned relation shard list
+ * contains shard intervals from multiple distributed tables.
+ */
+static bool
+MultipleDistributedTableList(List *relationShardLists)
+{
+	ListCell *relationShardListCell = NULL;
+	List *relationIds = NIL;
+
+	foreach(relationShardListCell, relationShardLists)
+	{
+		List *relationShardList = (List *) lfirst(relationShardListCell);
+		ListCell *shardIntervalCell = NULL;
+
+		foreach(shardIntervalCell, relationShardList)
+		{
+			ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+			Oid relationId = shardInterval->relationId;
+
+			relationIds = list_append_unique_oid(relationIds, relationId);
+
+			if (list_length(relationIds) > 1)
```

Comment:
Using a list here seems a little confusing. I think it would make more sense to declare an `Oid firstRelationId = InvalidOid;` variable, and then:

```C
if (firstRelationId == InvalidOid) 
{
    firstRelationId = relationId;
}
else if (relationId != firstRelationId)
{
    return true;
}
```